storage/listv2.go

```diff
@@ -0,0 +1,94 @@
+// Copyright (C) 2018 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package storage
+
+import "errors"
+
+// More indicates if the result was truncated. If false
+// then the result []ListItem includes all requested keys.
+// If true then the caller must call List again to get more
+// results by setting `StartAfter` or `EndBefore` appropriately.
+type More bool
+
+// ListOptions are items that are optional for the LIST method
+type ListOptions struct {
+	Prefix       Key
+	StartAfter   Key
+	EndBefore    Key
+	Recursive    bool
+	IncludeValue bool
+	Limit        Limit
+}
+
+// ListV2 lists all keys corresponding to ListOptions
+func ListV2(store KeyValueStore, opts ListOptions) (result Items, more More, err error) {
+	if opts.StartAfter != nil && opts.EndBefore != nil {
+		return nil, false, errors.New("start-after and end-before cannot be combined")
+	}
+
+	reverse := opts.EndBefore != nil
+
+	more = More(true)
+	limit := opts.Limit
+	if limit == 0 {
+		limit = Limit(1 << 31)
+	}
+
+	var first Key
+	if !reverse {
+		first = opts.StartAfter
+	} else {
+		first = opts.EndBefore
+	}
+
+	iterate := func(it Iterator) error {
+		var item ListItem
+		skipFirst := true
+		for ; limit > 0; limit-- {
+			if !it.Next(&item) {
+				more = false
+				return nil
+			}
+			if skipFirst {
+				skipFirst = false
+				if item.Key.Equal(first) {
```

Comment:
What if the `first` has been deleted just before making the call?

We need are more robust check. Instead of waiting to hit the `first` item and skip it, it would be better to check that the current item is "greater" than `opt.StartAfter` and "lower" than `opt.EndBefore`.

Reply:
Yes, you are right. The iterator handles this. I've just added a few more test cases to check for this too.